File: app/app/src/main/java/com/cloudcast/app/crypto/Sodium.kt
package com.cloudcast.app.crypto

import com.goterl.lazysodium.LazySodiumAndroid
import com.goterl.lazysodium.SodiumAndroid
import com.goterl.lazysodium.interfaces.Box
import com.goterl.lazysodium.interfaces.GenericHash
import com.goterl.lazysodium.interfaces.PwHash
import com.goterl.lazysodium.interfaces.DiffieHellman
import com.goterl.lazysodium.interfaces.SecretBox
import com.sun.jna.NativeLong

object Sodium {

    object Argon2id {
        const val OPS = 3L
        const val MEM = 67108864L
        const val SALT_BYTES = 16
        const val KEY_BYTES = 32
    }

    const val SEALED_BOX_OVERHEAD = 48
    const val SECRETBOX_NONCE_BYTES = 24
    const val SECRETBOX_MAC_BYTES = 16
    const val PUBLIC_KEY_BYTES = 32
    const val SECRET_KEY_BYTES = 32

    private val ls: LazySodiumAndroid by lazy { LazySodiumAndroid(SodiumAndroid()) }

    fun ready() {

        check(ls.randomBytesBuf(1).size == 1) { "libsodium failed to initialise" }
    }

    fun randomBytes(n: Int): ByteArray = ls.randomBytesBuf(n)
    fun generateDek(): ByteArray = randomBytes(32)
    fun generateSalt(): ByteArray = randomBytes(Argon2id.SALT_BYTES)

    data class Keypair(val publicKey: ByteArray, val privateKey: ByteArray) {

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Keypair) return false
            return publicKey.contentEquals(other.publicKey) &&
                privateKey.contentEquals(other.privateKey)
        }

        override fun hashCode(): Int = 31 * publicKey.contentHashCode() + privateKey.contentHashCode()
    }

    fun generateKeypair(): Keypair {
        val pk = ByteArray(PUBLIC_KEY_BYTES)
        val sk = ByteArray(SECRET_KEY_BYTES)
        check((ls as Box.Native).cryptoBoxKeypair(pk, sk)) { "crypto_box_keypair failed" }
        return Keypair(pk, sk)
    }

    fun keypairFromSeed(seed: ByteArray): Keypair {
        require(seed.size == 32) { "seed must be 32 bytes, was ${seed.size}" }
        val pk = ByteArray(PUBLIC_KEY_BYTES)
        val sk = ByteArray(SECRET_KEY_BYTES)
        check((ls as Box.Native).cryptoBoxSeedKeypair(pk, sk, seed)) { "crypto_box_seed_keypair failed" }
        return Keypair(pk, sk)
    }

    fun publicKeyFromSecret(secretKey: ByteArray): ByteArray {
        require(secretKey.size == SECRET_KEY_BYTES) { "secret key must be $SECRET_KEY_BYTES bytes" }
        val pk = ByteArray(PUBLIC_KEY_BYTES)
        check((ls as DiffieHellman.Native).cryptoScalarMultBase(pk, secretKey)) {
            "crypto_scalarmult_base failed"
        }
        return pk
    }

    fun deriveVaultKey(password: String, salt: ByteArray): ByteArray {
        require(salt.size == Argon2id.SALT_BYTES) { "salt must be ${Argon2id.SALT_BYTES} bytes" }
        val out = ByteArray(Argon2id.KEY_BYTES)
        val pw = password.toByteArray(Charsets.UTF_8)
        val ok = (ls as PwHash.Native).cryptoPwHash(
            out, out.size, pw, pw.size, salt,
            Argon2id.OPS, NativeLong(Argon2id.MEM), PwHash.Alg.PWHASH_ALG_ARGON2ID13,
        )
        pw.fill(0)
        check(ok) { "crypto_pwhash failed (out of memory?)" }
        return out
    }

    const val SRP_STRETCH_CONTEXT = "cloudcast-srp-v1:"

    /**
     * The SRP verifier is derived from the password, and a leaked database contains it.
     * Stretching the password with Argon2id first makes guessing against the verifier cost
     * the same as guessing against the vault, instead of a single SHA-256.
     */
    fun srpPassword(password: String, srpSaltHex: String): String {
        val context = (SRP_STRETCH_CONTEXT + srpSaltHex).toByteArray(Charsets.UTF_8)
        val salt = ByteArray(Argon2id.SALT_BYTES)
        check(
            (ls as GenericHash.Native).cryptoGenericHash(salt, salt.size, context, context.size.toLong(), null, 0),
        ) { "crypto_generichash failed" }
        val stretched = deriveVaultKey(password, salt)
        val hex = stretched.joinToString("") { "%02x".format(it) }
        stretched.wipe()
        return hex
    }

    fun recoveryVaultKey(recoveryKey: ByteArray): ByteArray {
        val out = ByteArray(32)
        check(
            (ls as GenericHash.Native).cryptoGenericHash(out, out.size, recoveryKey, recoveryKey.size.toLong(), null, 0),
        ) { "crypto_generichash failed" }
        return out
    }

    fun secretboxSeal(message: ByteArray, key: ByteArray): ByteArray =
        secretboxSealWithNonce(message, key, randomBytes(SECRETBOX_NONCE_BYTES))

    fun secretboxSealWithNonce(message: ByteArray, key: ByteArray, nonce: ByteArray): ByteArray {
        require(nonce.size == SECRETBOX_NONCE_BYTES) { "nonce must be $SECRETBOX_NONCE_BYTES bytes" }
        val ct = ByteArray(message.size + SECRETBOX_MAC_BYTES)
        check((ls as SecretBox.Native).cryptoSecretBoxEasy(ct, message, message.size.toLong(), nonce, key)) {
            "crypto_secretbox_easy failed"
        }
        return nonce + ct
    }

    fun secretboxOpen(blob: ByteArray, key: ByteArray): ByteArray {
        require(blob.size > SECRETBOX_NONCE_BYTES + SECRETBOX_MAC_BYTES) { "secretbox blob is truncated" }
        val nonce = blob.copyOfRange(0, SECRETBOX_NONCE_BYTES)
        val ct = blob.copyOfRange(SECRETBOX_NONCE_BYTES, blob.size)
        val out = ByteArray(ct.size - SECRETBOX_MAC_BYTES)
        check((ls as SecretBox.Native).cryptoSecretBoxOpenEasy(out, ct, ct.size.toLong(), nonce, key)) {
            "secretbox open failed - wrong key or corrupt data"
        }
        return out
    }

    fun sealTo(message: ByteArray, recipientPk: ByteArray): ByteArray {
        val out = ByteArray(message.size + SEALED_BOX_OVERHEAD)
        check((ls as Box.Native).cryptoBoxSeal(out, message, message.size.toLong(), recipientPk)) {
            "crypto_box_seal failed"
        }
        return out
    }

    fun sealOpen(blob: ByteArray, pk: ByteArray, sk: ByteArray): ByteArray {
        require(blob.size > SEALED_BOX_OVERHEAD) { "sealed box is truncated" }
        val out = ByteArray(blob.size - SEALED_BOX_OVERHEAD)
        check((ls as Box.Native).cryptoBoxSealOpen(out, blob, blob.size.toLong(), pk, sk)) {
            "sealed box open failed - not sealed to this key, or corrupt"
        }
        return out
    }
}

fun ByteArray.wipe() = fill(0)

File: web/src/features/crypto/sodium.ts
import _sodium from "libsodium-wrappers-sumo";

let s: typeof _sodium | null = null;
export async function ready(): Promise<void> {
  if (!s) { await _sodium.ready; s = _sodium; }
}
function S(): typeof _sodium {
  if (!s) throw new Error("crypto not initialised - await ready() first");
  return s;
}

export const ARGON2ID = { OPS: 3, MEM: 67108864, SALTBYTES: 16, KEYBYTES: 32 } as const;

export const toB64 = (b: Uint8Array): string => S().to_base64(b, S().base64_variants.ORIGINAL);
export const fromB64 = (t: string): Uint8Array => S().from_base64(t, S().base64_variants.ORIGINAL);
export const toB64Url = (b: Uint8Array): string => S().to_base64(b, S().base64_variants.URLSAFE_NO_PADDING);
export const fromB64Url = (t: string): Uint8Array => S().from_base64(t, S().base64_variants.URLSAFE_NO_PADDING);
export const toHex = (b: Uint8Array): string => S().to_hex(b);
export const fromHex = (t: string): Uint8Array => S().from_hex(t);
export const utf8 = (t: string): Uint8Array => S().from_string(t);
export const fromUtf8 = (b: Uint8Array): string => S().to_string(b);

export interface Keypair { publicKey: Uint8Array; privateKey: Uint8Array; }

export const randomBytes = (n: number): Uint8Array => S().randombytes_buf(n);
export const generateDek = (): Uint8Array => randomBytes(32);
export const generateSalt = (): Uint8Array => randomBytes(ARGON2ID.SALTBYTES);

export function generateKeypair(): Keypair {
  const kp = S().crypto_box_keypair();
  return { publicKey: kp.publicKey, privateKey: kp.privateKey };
}
export function keypairFromSeed(seed: Uint8Array): Keypair {
  const kp = S().crypto_box_seed_keypair(seed);
  return { publicKey: kp.publicKey, privateKey: kp.privateKey };
}
export function publicFromSecret(sk: Uint8Array): Uint8Array {
  return S().crypto_scalarmult_base(sk);
}

export function deriveVaultKey(password: string, salt: Uint8Array): Uint8Array {
  return S().crypto_pwhash(
    ARGON2ID.KEYBYTES, password, salt, ARGON2ID.OPS, ARGON2ID.MEM, S().crypto_pwhash_ALG_ARGON2ID13);
}
export const SRP_STRETCH_CONTEXT = "cloudcast-srp-v1:";

// The SRP verifier is derived from the password, and a leaked database contains it.
// Stretching the password with Argon2id first makes guessing against the verifier cost
// the same as guessing against the vault, instead of a single SHA-256.
export function srpPassword(password: string, srpSaltHex: string): string {
  const salt = S().crypto_generichash(ARGON2ID.SALTBYTES, utf8(SRP_STRETCH_CONTEXT + srpSaltHex));
  const stretched = deriveVaultKey(password, salt);
  const hex = toHex(stretched);
  wipe(stretched);
  return hex;
}

export function recoveryVaultKey(recoveryKey: Uint8Array): Uint8Array {
  return S().crypto_generichash(32, recoveryKey);
}

export function secretboxSeal(message: Uint8Array, key: Uint8Array): Uint8Array {
  const nonce = S().randombytes_buf(S().crypto_secretbox_NONCEBYTES);
  const ct = S().crypto_secretbox_easy(message, nonce, key);
  const out = new Uint8Array(nonce.length + ct.length);
  out.set(nonce); out.set(ct, nonce.length);
  return out;
}
export function secretboxSealWithNonce(message: Uint8Array, key: Uint8Array, nonce: Uint8Array): Uint8Array {
  const ct = S().crypto_secretbox_easy(message, nonce, key);
  const out = new Uint8Array(nonce.length + ct.length);
  out.set(nonce); out.set(ct, nonce.length);
  return out;
}

export function secretboxOpen(blob: Uint8Array, key: Uint8Array): Uint8Array {
  const n = S().crypto_secretbox_NONCEBYTES;
  return S().crypto_secretbox_open_easy(blob.slice(n), blob.slice(0, n), key);
}

export const sealTo = (message: Uint8Array, pk: Uint8Array): Uint8Array => S().crypto_box_seal(message, pk);
export const sealOpen = (blob: Uint8Array, pk: Uint8Array, sk: Uint8Array): Uint8Array => S().crypto_box_seal_open(blob, pk, sk);

export function wipe(...buffers: (Uint8Array | undefined | null)[]): void {
  for (const b of buffers) if (b) b.fill(0);
}
